pkgs/_test_macros/lib/json_codable.dart

```diff
@@ -0,0 +1,170 @@
+      runsInPhases: [2, 3]);
+
+  @override
+  Future<AugmentResponse> augment(Host host, AugmentRequest request) async {
+    return switch (request.phase) {
+      1 => phase1(host, request),
+      2 => phase2(host, request),
+      _ => AugmentResponse(augmentations: []),
+    };
+  }
+
+  Future<AugmentResponse> phase1(Host host, AugmentRequest request) async {
+    final target = request.target;
+    return AugmentResponse(augmentations: [
+      Augmentation(code: '''
+external ${target.code}.fromJson($_jsonMapType json);
+external $_jsonMapType toJson();
+   '''),
+    ]);
+  }
+
+  Future<AugmentResponse> phase2(Host host, AugmentRequest request) async {
+    final result = <Augmentation>[];
+
+    final target = request.target;
+    final model = await host.query(Query(target: target));
+    final clazz = model.uris[target.uri]!.scopes[target.name]!;
+    final initializers = <String>[];
+    for (final field
+        in clazz.members.entries.where((m) => m.value.properties.isField)) {
+      final name = field.key;
+      final type = field.value.returnType;
+      initializers
+          .add('$name = ${_convertTypeFromJson("json[r'$name']", type)}');
+    }
+
+    // TODO(davidmorgan): helper for augmenting initializers.
+    result.add(Augmentation(code: '''
+augment ${target.code}.fromJson($_jsonMapType json) :
+${initializers.join(',\n')};
+'''));
+
+    final serializers = <String>[];
+    for (final field
+        in clazz.members.entries.where((m) => m.value.properties.isField)) {
+      final name = field.key;
+      final type = field.value.returnType;
+      serializers.add("json[r'$name'] = ${_convertTypeToJson(name, type)}");
+    }
+
+    // TODO(davidmorgan): helper for augmenting methods.
+    result.add(Augmentation(code: '''
+$_jsonMapType toJson() {
+${serializers.join(';\n')}
+};
+'''));
+
+    return AugmentResponse(augmentations: result);
+  }
+
+  String _convertTypeFromJson(String reference, StaticTypeDesc type) {
+    // TODO(davidmorgan): _checkNamedType equivalent.
+    final nullable = type.type == StaticTypeDescType.nullableTypeDesc;
+    final orNull = nullable ? '?' : '';
+    final nullCheck = nullable ? '' : '$reference == null ? null : ';
+    final underlyingType = type.type == StaticTypeDescType.nullableTypeDesc
+        ? type.asNullableTypeDesc.inner
+        : type;
+
+    if (underlyingType.type == StaticTypeDescType.namedTypeDesc) {
+      final namedType = underlyingType.asNamedTypeDesc;
+      if (namedType.name.uri == 'dart:core') {
+        switch (namedType.name.name) {
+          case 'bool':
+          case 'String':
+          case 'int':
+          case 'double':
+          case 'num':
+            return '$reference as ${namedType.name.code}$orNull';
+          case 'List':
+            return '$nullCheck [for (final item in $reference '
+                'as {{dart:core#List}}<{{dart:core#Object}}?>) '
+                '${_convertTypeFromJson('item', namedType.instantiation.first)}'
```

Comment:
Should this really be `.single`? Presumably these are the type arguments? We probably want to rename `instantiation` to `typeArguments` as well.

Reply:
We probably also need a check that there were any type arguments provided (I believe the existing macro does this).